x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/state.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient, Logger } from '@kbn/core/server';
+import { StreamsStorageClient } from '../service';
+import { StreamActiveRecord, StreamChange, ValidationResult } from './types';
+import { streamFromDefinition } from './stream_from_definition';
+
+export class State {
+  private streamsByName: Map<string, StreamActiveRecord>;
+
+  constructor(streams: StreamActiveRecord[]) {
+    this.streamsByName = new Map();
+    streams.forEach((stream) => this.streamsByName.set(stream.definition.name, stream));
+  }
+
+  applyChanges(requestedChanges: StreamChange[]): State {
+    // What if multiple changes target the same stream?
+
+    // Optional: Expand requested changes to include automatic creation of missing streams
+    const newState = this.clone();
+
+    requestedChanges.forEach((change) => {
```

Comment:
I'm wondering if this process shouldn't be more similar to the validation step and return a result like `{ changedSuccessfully: boolean; errors: Error[] }` so that it's possible to reject changes that cannot be applied even outside of validation (and then wrap the calls to `update` / `markForDeletion` in a try/catch so that this part never throw any unhandled exceptions).